### ISSUE TITLE
[Feature Request] How to build Racket for ALT Linux

### ISSUE THREAD

**Issue by Ancieg** [racket/racket#4456]
**Is your feature request related to a problem? Please describe.**
We build packages in ALT Linux via `hasher`, the program which creates isolated environment during building (chroot). That building environment are strictly disconnected from the Internet (for some reasons, like a security, more clean result and so on).
We use rpm as a package management tool.

I need to describe our build process briefly:
1. Hasher creates isolated environment and copy source files (tar archive) into `/usr/src/RPM/SOURCES` in chroot (spec file is copied to `/usr/src/RPM/SPECS`).
2. Hasher installs all needed packages to build (they are taken from BuildRequires). The following steps are all in the chroot.
3. RPM starts building using spec from `/usr/src/RPM/SPECS`.
3.1. **(%prep)** `/usr/src/RPM/SOURCES/<name>-<version>.tar` is extracted to `/usr/src/RPM/BUILD/<name>-<version>/`
3.2. **(%build)** The program is built in-place (or into subdirectory).
3.3. **(%install)** Move building results into `%buildroot` (in our case it is `/usr/src/tmp/<name>-buildroot/`). In `%buildroot` we create paths like they are related to `/`: for example, `/usr/bin/<program-name>`. But full file path is
`/usr/src/tmp/<program-name>-buildroot/usr/bin/<program-name>`.
3.4. **(%files)** We specify files to package relative to `%buildroot`, `/usr/bin/<name>` by example (it will be moved to RPM-package from `%buildroot/usr/bin/<name>`).
4. Creates RPM package with files taken from **%files** section.
5. We install the package into the system (without `%buildroot` prefix in files paths).

The problem is that Racket has hard-coded paths inside binaries and libraries, that is `/usr/src/tmp/<program>-buildroot/` prefixes all paths, but it doesn't exist in user's system (in non-chroot).

**Describe the solution you'd like**
I want to know how to build and install Racket via way described above so that the paths do not contain %buildroot as a prefix.

As an alternative, I would like to see generation more flexible Makefiles for the Racket source code repository.

**Do you want to contribute to this feature**
Yes, I want. But I can't because I novice to Racket and learn "How to design programs" book. But I can't deep into Racket learning, because we don't have Racket in our repository (Sisyphus) for now (previous maintainer can't maintain for now, so I want to build the package for our repository).

---

(I apologize for my poor English language skills)

**Comment by LiberalArtist** [comment]
Can you point me toward the build script or package definition your ALT Linux package is currently trying to use?

It definitely should be possible to do this. From the top level of the Git reposirory, I know you can do this:
https://github.com/racket/racket/blob/971c04d0798b33d99c5ff090934384342342598b/build.md?plain=1#L159-L163

I thought I remembered a `configure` option equivalent to `DESTDIR` (to be used in conjunction with `configure --prefix=/usr`), but I didn't find it just now. Maybe I misremembered, or possibly it could have been lost in the transition to Zuo.

(It's nice to see a descendant of Mandrake Linux: that was the first distribution I could get to run X successfully, way back when.)

**Comment by Ancieg** [comment]
> Can you point me toward the build script or package definition your ALT Linux package is currently trying to use?

You can see our robot reported for the last version here (a lot of %buildroot's found in paths): https://packages.altlinux.org/en/sisyphus/srpms/racket/repocop/
Spec: https://packages.altlinux.org/en/sisyphus/srpms/racket/specfiles/

**Comment by LiberalArtist** [comment]
I'm not fluent in rpm specfiles, but where you currently have:
```
%makeinstall_std -j1 -C src docdir=%_docdir/%name-%version

install -p -m644 README %buildroot%_docdir/%name-%version/
sed -i 's|%buildroot||g' %buildroot%_desktopdir/*.desktop
```

I believe you need to add `DESTDIR=%buildroot%` to the call to `%makeinstall_std%`. (Apparently that's typical for Autoconf projects, rather than using a `configure` flag.) I would not expect you to need to pass `docdir=%_docdir/%name-%version`, since you supply `--docdir=%_docdir/%name-%version` to `configure`.

I think the call to `sed` should be unnecessary if things are working properly.

There are other things you might change in the `configure` flags—for instance, `--enable-shared` has no effect with Racket CS, and you can now build CS for ppc64le via the "portable bytecode" backends—but I think `DESTDIR` is the main missing piece.

**Comment by Ancieg** [comment]
`%makeinstall_std` is expanded to `make INSTALL="/bin/install -p" install DESTDIR=%buildroot'` (`%buildroot` -> `/usr/src/tmp/racket-buildroot`).

**Comment by Ancieg** [comment]
The Makefile from the top level of the Git repository make `raco` to download pkgs from the Internet (https://download.racket-lang.org/releases/8.6/catalog/), but our building chroot doesn't allow it:
```
raco setup: --- checking package dependencies ---                  [15:52:14]
/usr/src/tmp/racket-buildroot/bin/raco pkg config -i --set catalogs https://download.racket-lang.org/releases/8.6/catalog/ ""
/usr/src/tmp/racket-buildroot/bin/racket -l- pkg/dirs-catalog --check-metadata build/local/pkgs-catalog pkgs
Finding packages
/usr/src/tmp/racket-buildroot/bin/raco pkg catalog-copy --force --from-config build/local/pkgs-catalog build/local/catalog
tcp-connect: host not found
  hostname: download.racket-lang.org
  port number: 443
  system error: Temporary failure in name resolution; gai_err=-3
  context...:
   /usr/src/tmp/racket-buildroot/share/racket/collects/openssl/mzssl.rkt:1940:0: do-ssl-connect
   /usr/src/tmp/racket-buildroot/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/src/tmp/racket-buildroot/share/racket/collects/net/http-client.rkt:67:0: http-conn-open!
   /usr/src/tmp/racket-buildroot/share/racket/collects/net/http-client.rkt:274:0: http-conn-open
   /usr/src/tmp/racket-buildroot/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/src/tmp/racket-buildroot/share/racket/collects/net/url.rkt:202:0: http://getpost-impure-port
   /usr/src/tmp/racket-buildroot/share/racket/collects/net/url.rkt:305:0: get-pure-port/headers
   /usr/src/tmp/racket-buildroot/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/src/tmp/racket-buildroot/share/racket/collects/pkg/private/network.rkt:59:3
   /usr/src/tmp/racket-buildroot/share/racket/collects/pkg/private/catalog.rkt:218:0: read-from-server
   /usr/src/tmp/racket-buildroot/share/racket/collects/pkg/private/catalog.rkt:298:0: get-all-pkg-details-from-catalogs
   /usr/src/tmp/racket-buildroot/share/racket/collects/pkg/private/catalog-copy.rkt:33:0: pkg-catalog-copy
   /usr/src/tmp/racket-buildroot/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   body of (submod "/usr/src/tmp/racket-buildroot/share/racket/collects/pkg/main.rkt" main)
   /usr/src/tmp/racket-buildroot/share/racket/collects/raco/raco.rkt:41:0
   body of "/usr/src/tmp/racket-buildroot/share/racket/collects/raco/raco.rkt"
   ...
failed
 in local-catalog
 in unix-style {2}
 in build-one
 in loop
make: *** [Makefile:198: unix-style] Error 1
```
Is there any way to build and install Racket with pkgs from the downloaded repository? There are the following packages:
```
$ ls pkgs
at-exp-lib  compiler      compiler-test  net-doc  net-test      racket-benchmarks   racket-doc    racket-lib   racket-test-core   sandbox-lib
base        compiler-lib  net            net-lib  plt-services  racket-build-guide  racket-index  racket-test  racket-test-extra  zo-lib
```
 I tried to find options to do that in the Makefile from the top level of the Git repository, but found nothing.

**Comment by Ancieg** [comment]
If I install Racket by the following command (after `make base`):
```
$ make unix-style DESTDIR=/usr/src/tmp/racket-buildroot/ PREFIX=/usr
```
I get the following path in `/usr/src/tmp/racket-buildroot/bin/raco` file:
```
# {{{ bindir
bindir="/usr/src/tmp/racket-buildroot/usr/bin"
# }}} bindir
```
but expectation is:
```
# {{{ bindir
bindir="/usr/bin"
# }}} bindir
```

**Comment by LiberalArtist** [comment]
> Is there any way to build and install Racket with pkgs from the downloaded repository? There are the following packages:
> 
> ```
> $ ls pkgs
> at-exp-lib  compiler      compiler-test  net-doc  net-test      racket-benchmarks   racket-doc    racket-lib   racket-test-core   sandbox-lib
> base        compiler-lib  net            net-lib  plt-services  racket-build-guide  racket-index  racket-test  racket-test-extra  zo-lib
> ```
> 
> I tried to find options to do that in the Makefile from the top level of the Git repository, but found nothing.

I'd suggest not using the top-level Makefile for distro packaging: it is oriented toward development and cooperation with [`distro-build`](https://docs.racket-lang.org/distro-build/index.html) to produce release tarballs. It is possible to build for a distro from the Git repository using `configure` and `make` in the [racket/src/](https://github.com/racket/racket/tree/master/racket/src) directory (I do this for the Guix package of Racket), but it means you have to get the Racket packages in `main-distribution` in some other way.

Every distribution I'm aware of but Guix builds from the release tarballs, and it should be possible to make it work with `DESTDIR=`: if the build scripts need to change to make it work, they should.

Do you have a log from a failing build of the ALT Linux package? It might be easier for me understand what actually happens than to try to work out from the RPM spec what will happen.

**Comment by Ancieg** [comment]
> Every distribution I'm aware of but Guix builds from the release tarballs, and it should be possible to make it work with `DESTDIR=`: if the build scripts need to change to make it work, they should.

Yes, that works properly. But I have faced problems on PPC64LE and ARM platforms (the builds on these platforms fail for unclear reasons). So I have decided to change the approach for building: I am trying to build ChezScheme first, but facing problems too. You can see building log on PPC64LE here: https://git.altlinux.org/tasks/308500/build/200/ppc64le/log

> Do you have a log from a failing build of the ALT Linux package? It might be easier for me understand what actually happens than to try to work out from the RPM spec what will happen.

See the reference above.

**Comment by Ancieg** [comment]
You can also see logs here (there I tries to build Racket base): https://git.altlinux.org/tasks/308520/ (choose arch then click log).

**Comment by samth** [comment]
The problem with armh seems to be that the target triple includes more - than the parsing expects.

The 2 hour build time for ppc64 is not that surprising unless there's new hardware for that platform recently.

**Comment by Ancieg** [comment]
> The problem with armh seems to be that the target triple includes more - than the parsing expects.

We have the following targets triples: `aarch64-alt-linux-gnu`, `armh-alt-linux-gnueabi`, `i586-alt-linux-gnu`, `powerpc64le-alt-linux-gnu` and `x86_64-alt-linux-gnu`.
I tried to remove `--build` and `--host` for `configure` and faced another problem on armh (builds on all of the our platform still work, except armh): https://git.altlinux.org/tasks/308554/build/200/armh/log.
```
[00:44:36] Converting cs/c/racket.boot to racket-pbchunk.boot and racket~a.c
[00:50:38] Exception: invalid memory reference.  Some debugging context lost
[00:50:38] failed
[00:50:38]  in build-one
[00:50:38]  in loop
[00:50:38]  in module->hash
```

**Comment by Ancieg** [comment]
:upside_down_face: I have just remove usage of portable bytecode for building on armh and it has been built successfully.
See: https://git.altlinux.org/tasks/308560/

**Comment by Ancieg** [comment]
Is there a way to install a package via `raco` using the same logic as `DESTDIR=`?
I tried to use `raco pkg install --scope-dir %{buildroot}/usr/lib/racket/pkgs/ --copy --type dir [pkg dir]`, but it doesn't compile packages (+ it doesn't create launchers, docs and so on).

**Comment by Ancieg** [comment]
It would be great to make `raco` follow the configuration file.
Then the command `PLTCONFIGDIR=%{buildroot}/etc/racket/ raco pkg install --copy -i -t dir *` with overrided paths in `config.rktd` will do what I need (at least I think so).

**Comment by samth** [comment]
You can do this with `racket -G %{buildroot}/etc/racket/ -l raco -- pkg install --copy -i -t dir *`

**Comment by Ancieg** [comment]
> You can do this with `racket -G %{buildroot}/etc/racket/ -l raco -- pkg install --copy -i -t dir *`

Looks like `raco` ignores `-G` option:
```
[builder@localhost pkgs]$ racket -G /usr/src/tmp/racket-main-distribution-buildroot/etc/racket/ -l raco -- pkg install --copy -i -t dir *
make-directory: cannot make directory
  path: /usr/lib/racket/share/
  system error: Permission denied; errno=13
  context...:
   /usr/lib/racket/collects/racket/file.rkt:134:0: make-directory*
   /usr/lib/racket/collects/pkg/private/lock.rkt:26:0: with-pkg-lock*
   /usr/lib/racket/collects/pkg/main.rkt:216:16
   body of (submod "/usr/lib/racket/collects/pkg/main.rkt" main)
   /usr/lib/racket/collects/raco/raco.rkt:41:0
   body of "/usr/lib/racket/collects/raco/raco.rkt"
   body of "/usr/lib/racket/collects/raco/main.rkt"
```

**Comment by samth** [comment]
You may need to set the addon-dir with `-A` as well.

**Comment by Ancieg** [comment]
> You may need to set the addon-dir with `-A` as well.

`raco` still ignores `%buildroot` (`/usr/src/tmp/racket-main-distribution-buildroot/`) path:
```
[builder@localhost pkgs]$ racket -A /usr/src/tmp/racket-main-distribution-buildroot/etc/racket -l raco -- pkg install --copy -i -t dir *
make-directory: cannot make directory
  path: /usr/lib/racket/pkgs
  system error: Permission denied; errno=13
  context...:
   /usr/lib/racket/collects/pkg/private/lock.rkt:26:0: with-pkg-lock*
   /usr/lib/racket/collects/pkg/main.rkt:216:16
   body of (submod "/usr/lib/racket/collects/pkg/main.rkt" main)
   /usr/lib/racket/collects/raco/raco.rkt:41:0
   body of "/usr/lib/racket/collects/raco/raco.rkt"
   body of "/usr/lib/racket/collects/raco/main.rkt"
```

File `/usr/src/tmp/racket-main-distribution-buildroot/etc/racket/config.rkt` contains the following:
```
#hash(
      (doc-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/share/doc/racket")
      (lib-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/lib64/racket")
      (pkgs-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/lib/racket/pkgs")
      (share-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/share/racket")
      (include-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/include/racket")
      (bin-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/bin")
      (apps-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/share/applications")
      (man-dir . "/usr/src/tmp/racket-main-distribution-buildroot/usr/share/man")
      (absolute-installation? . #t)
)
```

**Comment by mflatt** [comment]
> If I install Racket by the following command (after `make base`):
> `$ make unix-style DESTDIR=/usr/src/tmp/racket-buildroot/ PREFIX=/usr`

I'm not sure that it works to use`make base` plus `make unix-style`. The `unix-style` target is meant to be used by itself, and `make base` may create a build that is incompatible with a later use of `make unix-style`. When I tried it just now, it worked better than I expected, but I don't trust it to work right.

It looks like you may also be trying to use the installation put into `DESTDIR` before it is moved into its final location. If so, that's going to be another source of trouble, since it's not meant to work that way. The intent is that after a build is put into `DESTDIR`, it's not used again until being moved to its final location. But if I understand right, probably this is already clear to you.

A `DESTDIR` build mostly works like a normal build, but there's a fixup step implemented by `setup/unixstyle-install`. It's possible that you want to run that script directly, perhaps twice: once after using `DESTDIR` and to put things back to where you can install more packages, and then again after the packages are installed to get back to a state that is ready for installation. I'm not sure that will work, but it might be in the right direction.

The addon directory should not be relevant. I'm not sure about the approach of temporarily redirecting via `config.rktd`; in principle, that approach can do a lot, but it's easy to get lost in the layers of search paths. Also, paths from a config file at build time do get copied into built artifacts like launcher scripts, and that's the kind of thing that `setup/unixstyle-install` fixes up.

**Comment by Ancieg** [comment]
I found a hacky way to do what I need:
```
fakechroot -s chroot . /bin/bash -c '/usr/bin/racket -l raco/main.rkt pkg install --skip-installed --copy -i -t dir /tmp/pkgs/*'
```
(I copied all needed files to the working directory (which is `%buildroot`): bash, racket, raco, collects...)
It looks like a working way. I am trying to use it now. Will report you the results soon.

**Comment by Ancieg** [comment]
> I found a hacky way to do what I need:
> 
> ```
> fakechroot -s chroot . /bin/bash -c '/usr/bin/racket -l raco/main.rkt pkg install --skip-installed --copy -i -t dir /tmp/pkgs/*'
> ```
> 
> (I copied all needed files to the working directory (which is `%buildroot`): bash, racket, raco, collects...) It looks like a working way. I am trying to use it now. Will report you the results soon.

This way works perfectly, but I encountered a problem within our building environment on `i586`, `ppc64le` and `armh`, example:
```
[00:43:19] raco setup: 5 making: <pkgs>/profile-doc/profile/scribblings
[00:43:30] raco setup: 5 making: <pkgs>/profile-lib
[00:43:32] raco setup: 5 making: <pkgs>/quickscript (Quickscript)
[00:43:41] /usr/sbin/chroot.fakechroot: line 147: 2435617 CPU time limit exceeded env -u FAKECHROOT_BASE_ORIG FAKECHROOT_CMD_ORIG= LD_LIBRARY_PATH="$fakechroot_chroot_paths" FAKECHROOT_BASE="$fakechroot_chroot_base" "$fakechroot_chroot_chroot" "${@:1:$(($fakechroot_chroot_n - 1))}" "$fakechroot_chroot_final_newroot" "${@:$(($fakechroot_chroot_n + 1))}"
[00:43:41] error: Bad exit status from /usr/src/tmp/rpm-tmp.89941 (%build)
[00:43:41] 
[00:43:41] 
[00:43:41] RPM build errors:
[00:43:41]     Bad exit status from /usr/src/tmp/rpm-tmp.89941 (%build)
[00:43:41] Command exited with non-zero status 1
[00:43:41] 10388.92user 414.04system 43:33.32elapsed 413%CPU (0avgtext+0avgdata 2628288maxresident)k
[00:43:41] 0inputs+0outputs (0major+58367minor)pagefaults 0swaps
[00:43:41] hsh-rebuild: rebuild of `racket-main-distribution-8.6-alt1.src.rpm' failed.
```
This bug is not related to the Racket.

---

Also I have an issue with `raco pkg show`: it doesn't show installation-wide packages. I think it is related to removing `pkgs.rktd` from a directory which contains packages, but I am not sure (I decided to remove that file because it contain paths which don't exist after building).

**Comment by Ancieg** [comment]
Is there a way to decrease CPU eating during `raco pkg install`?

**Comment by sorawee** [comment]
Perhaps try `-j 1` to disable parallel jobs.

**Comment by Ancieg** [comment]
> Perhaps try `-j 1` to disable parallel jobs.

I will try using `-j 1`, but I am not really sure it will decrease CPU-time eating (our builder has limited value of `ulimit -t`).

**Comment by samth** [comment]
You should consider building fewer packages; compiling everything in the full Racket distribution is a lot of code and will take a long time.

**Comment by Ancieg** [comment]
> You should consider building fewer packages; compiling everything in the full Racket distribution is a lot of code and will take a long time.

Trying this now, but I'm suffering from the huge dependency tree of the main-distribution.
I would be glad if you tell me how to split the main-distribution into fewer parts.

**Comment by Ancieg** [comment]
> > You should consider building fewer packages; compiling everything in the full Racket distribution is a lot of code and will take a long time.
> 
> Trying this now, but I'm suffering from the huge dependency tree of the main-distribution. I would be glad if you tell me how to split the main-distribution into fewer parts.

Btw, is there an image of dependency tree of the main-distribution? I think it will help to split it into fewer parts.

**Comment by samth** [comment]
It depends what packages you want. I would start with just "base" as a package, which is truly minimal; just enough for people to install more packages. Other possibilities are "drracket", which is everything needed for the IDE, and "web-server" which is everything needed for the web.

**Comment by Ancieg** [comment]
In this issue I want to ask you how to handle such files as `launchers.rktd`, `links.rktd`, `mans.rktd` and so on.
The problem is related to file conflicts (when two or more different packages contain the same file with different content).
I see that there is a `raco link -i --repair` and I think that can resolve the problem with `links.rktd` file.
But what about `launchers.rktd`, `mans.rktd` and others?

**Comment by Ancieg** [comment]
> In this issue I want to ask you how to handle such files as `launchers.rktd`, `links.rktd`, `mans.rktd` and so on. The problem is related to file conflicts (when two or more different packages contain the same file with different content). I see that there is a `raco link -i --repair` and I think that can resolve the problem with `links.rktd` file. But what about `launchers.rktd`, `mans.rktd` and others?

I found the solution for that problem (via `RPMFileTrigger` and `.d`-approach).

---

Is there a way to find files required by an another `.rkt`-file?
I found [show-dependecies](https://docs.racket-lang.org/macro-debugger/index.html#%28def._%28%28lib._macro-debugger%2Fanalysis%2Fshow-dependencies..rkt%29._show-dependencies%29%29) procedure, but it doesn't fit my needs: I need only direct dependencies (not all).
Previously I tried to parse `.rkt`-files via `awk` and `sed`, but this approach has **many** problems.

**Comment by Ancieg** [comment]
`~/rkt/rpm-build-racket/moddep.rkt` (slightly modified `collects/syntax/moddep.rkt` module):
```
(require syntax/modread
         syntax/modcode
         syntax/modcollapse
         syntax/modresolve)

(define (show-import-tree module-path
                          #:dag? [dag? #f]
                          #:path-to [given-path-to #f]
                          #:show [show (lambda (indent path require-mode phase)
                                         (printf "~a~a~a ~a\n" indent path require-mode phase))])
  (define path-to (and given-path-to (simplify-path (resolve-module-path given-path-to #f))))
  (define seen (and dag? (make-hash)))
  (let loop ([path (resolve-module-path module-path #f)] [indent ""] [fs ""] [phase 0] [accum '()])
    (when (equal? path-to path)
      (let ([accum (let loop ([accum (cons (list indent path fs phase) accum)])
                     (cond
                       [(null? accum) null]
                       [(and seen (hash-ref seen accum #f)) null]
                       [else
                        (when seen (hash-set! seen accum #t))
                        (cons (car accum) (loop (cdr accum)))]))])
        (for ([i (in-list (reverse accum))])
          (apply show i))))
    (unless (and seen (hash-ref seen (cons path phase) #f))
      (unless path-to
        (show indent path fs phase))
      (when seen (hash-set! seen (cons path phase) #t))
      (define plain-path (if (pair? path) (cadr path) path))
      (let ([code (get-module-code plain-path
                                   #:submodule-path (if (pair? path) (cddr path) '()))])
        (let ([imports (module-compiled-imports code)]
              [accum (cons (list indent path fs phase) accum)])
          (define ((mk-loop phase-shift fs) i)
            (let ([p (resolve-module-path-index i plain-path)])
              (unless (symbol? p)
                (loop (if (path? p)
                          (simplify-path p)
                          (list* 'submod (simplify-path (cadr p)) (cddr p)))
                      (format " ~a" indent)
                      fs
                      (and phase phase-shift (+ phase phase-shift))
                      accum))))
          (for-each (lambda (i)
                      (for-each
                       (mk-loop (car i)
                                (case (car i)
                                  [(0) ""]
                                  [(1) " [for-syntax]"]
                                  [(-1) " [for-template]"]
                                  [(#f) " [for-label]"]
                                  [else (format " [for-meta ~a]" (car i))]))
                       (cdr i)))
                    imports))))))
```
```
$ cd /usr/lib/racket/pkgs && racket -f ~/rkt/rpm-build-racket/moddep.rkt -e "(show-import-tree \"2d/info.rkt\" #:dag? #t)" | sed -e 's/[[:digit:]]*$//' -e 's/\[.*\]//' | sed -E 's@/usr/lib/racket/(collects|pkgs)/@@' | sed -e 's/\.rkt[[:space:]]*$//' | xargs printf "racket(%s)\n" | sort -u
```
I think that would work (find `.rkt`-requires for `pkgs/2d/info.rkt` by example).

**Comment by LiberalArtist** [comment]
I stumbled today upon https://github.com/racket/racket/issues/3726, in which Fedora had similar-seeming problems with the build root, but maybe also false positives (or maybe none of this is useful; I haven't looked that closely). In particular, from https://github.com/racket/racket/issues/3726#issuecomment-798734769 it sounds like you may need to add something like `%global _find_debuginfo_opts --keep-section .rackboot` to your specfile.

**Comment by LiberalArtist** [comment]
> > > Perhaps try `-j 1` to disable parallel jobs.
> > 
> > 
> > I will try using `-j 1`, but I am not really sure it will decrease CPU-time eating (our builder has limited value of `ulimit -t`).
> 
> That helps building on `i586` only (so, now I get builds on `x86_64`, `aarch64` and `i586`). Builds on `ppc64le` and `armh` still crash due to `CPU time limit exceeded`.

Guix builds Racket CS for `ppc64le-linux` using the `tpb64l` backed. You can see the logs for the various steps in our bootstrap chain: [`zuo`](https://ci.guix.gnu.org/build/1329891/details), [`racket-vm-cgc`](https://ci.guix.gnu.org/build/1329890/details), [`racket-vm-bc`](https://ci.guix.gnu.org/build/1329893/details), [`chez-scheme-for-racket-bootstrap-bootfiles`](https://ci.guix.gnu.org/build/1329915/details), [`chez-scheme-for-racket`](https://ci.guix.gnu.org/build/1329919/details), [`racket-vm-cs`](https://ci.guix.gnu.org/build/1329892/details), [`racket-minimal`](https://ci.guix.gnu.org/build/1329894/details), and [`racket`](https://ci.guix.gnu.org/build/1329889/details) (the main distribution). It looks like the build took about 6 hours and 40 minutes (`(+ 384 1237 678 2890 3112 13670 57 16323)` seconds), and that is with the Chez Scheme test suite disabled.

Our armhf build machines have had issues lately, so I don't have numbers there. Even on x86_64-linux, though, the full build does take a while.

(Aside: @mflatt, I was surprised just now to see that building `racket-vm-cs` with pre-built Chez and Racket BC took 13,670 seconds, whereas the total time up to that point was 8,301, including for bootstrapping Chez via CGC. Is that sort of timing expected, maybe due to WPO or something? Or does it suggest that I may not be supplying the pre-built parts properly, or perhaps that `tpb64l` is not performing as well as BC? Of course, it could also just be some weirdness for that particular build: I can investigate further if it seems warranted.)

> Is there a way to find files required by an another `.rkt`-file?
> I found [show-dependecies](https://docs.racket-lang.org/macro-debugger/index.html#%28def._%28%28lib._macro-debugger%2Fanalysis%2Fshow-dependencies..rkt%29._show-dependencies%29%29) procedure, but it doesn't fit my needs: I need only direct dependencies (not all).
> Previously I tried to parse `.rkt`-files via `awk` and `sed`, but this approach has **many** problems.

I think `module-recorded-dependencies` from [`compiler/depend`](https://docs.racket-lang.org/raco/API_for_Reading_Compilation_Dependencies.html) might do what you want.

> > In this issue I want to ask you how to handle such files as `launchers.rktd`, `links.rktd`, `mans.rktd` and so on. The problem is related to file conflicts (when two or more different packages contain the same file with different content). I see that there is a `raco link -i --repair` and I think that can resolve the problem with `links.rktd` file. But what about `launchers.rktd`, `mans.rktd` and others?
> 
> I found the solution for that problem (via `RPMFileTrigger` and `.d`-approach).

I would strongly suggest that you avoid going down this route.

As you might gather from the build logs above, with the Guix packaging of Racket, I have gone pretty far down the road of splitting out pieces into separate Guix packages. No other distro packages Racket this way. I've done this in Guix for a few reasons, but in particular:

- We package a number of Chez Scheme libraries (e.g. https://github.com/fedeinthemix/chez-scmutils), and we build them against Racket's Chez Scheme variant on platforms not supported by upstream Chez; and
- Most significantly, I've been working to add support for Racket packages to Guix analogous to Guix's support for package systems like PyPI, NPM, Hackage, Ruby Gems, Crates.io, Elm, and many others. Guix distributes many packages from those systems, well beyond what other distros consider to be in scope, and Guix tools can automatically generate Guix package definitions for arbitrary packages from those systems, so users can import any additional packages they might need. So, my goal is to split our `racket` package into about 203 separate packages.

That is not at all what rpm- or deb-based distros want. It's quite different event from Gentoo or Arch, AIUI.

Other than Guix, distros have a `racket` package for the main distribution. A few distros also have something like Guix's `racket-minimal` package or the "Minimal Racket" installers from https://download.racket-lang.org/releases/8.6/, i.e. just the `base` package. (Nix does have a package they call, confusingly IMO, `chez-racket` analogous to Guix's `chez-scheme-for-racket`: IIUC the primary motivation is Idris2 on aarch64.)

I believe Guix is also the only distro to build from Git rather than release tarballs like https://download.racket-lang.org/releases/8.6/installers/racket-src.tgz, and even we don't use the top-level Makefile.

If support for building that way with `DESTDIR` is broken or has regressed, we (i.e. the Racket community) should fix it.

If builds on `ppc64le` and `armh` still exceed your distro's CPU time limit, I expect you will ultimately need to check with whoever makes decisions about such policies in your distro. It could make sense to make a `racket-minimal` package that `racket` would depend on. There may be other mitigations: IIRC Debian builds the documentation only once and uses it on architectures where building it is too intensive, and cross-compiling Racket code is possible technically if your distro's policy allows it. But I suspect that splitting up a package to get more CPU time would not be most distros' preferred solution.

If, despite all of that, you decide you do want to split Racket into multiple packages, I would still strongly advise against trying to manage `launchers.rktd`, `links.rktd`, `mans.rktd` manually. Instead, I'd suggest [building](https://docs.racket-lang.org/pkg/strip.html#%28tech._built._package%29) the packages you need and then installing them with `raco pkg`. I can offer some more detailed thoughts if you end up going this way.

> I found paths to `%buildroot` too using `less /usr/bin/racket`:
> 
> ```
> 000ca850  00 00 00 00 00 00 00 00  63 6f 4e 46 49 67 20 64  |........coNFIg d|
> 000ca860  49 52 45 43 54 4f 52 79  3a 2f 65 74 63 2f 72 61  |IRECTORy:/etc/ra|
> 000ca870  63 6b 65 74 00 00 72 61  63 6b 65 74 2d 62 61 73  |cket..racket-bas|
> 000ca880  65 2d 62 75 69 6c 64 72  6f 6f 74 2f 65 74 63 2f  |e-buildroot/etc/|
> 000ca890  72 61 63 6b 65 74 00 00  2a 2a 2a 2a 2a 2a 2a 2a  |racket..********|
> 000ca8a0  2a 2a 2a 2a 2a 2a 2a 2a  2a 2a 2a 2a 2a 2a 2a 2a  |****************|
> *
> 000cac90  2a 2a 2a 2a 2a 2a 2a 00  63 6f 4c 4c 45 43 54 73  |*******.coLLECTs|
> 000caca0  20 64 49 52 45 43 54 4f  52 79 3a 2f 75 73 72 2f  | dIRECTORy:/usr/|
> 000cacb0  6c 69 62 2f 72 61 63 6b  65 74 2f 63 6f 6c 6c 65  |lib/racket/colle|
> 000cacc0  63 74 73 00 00 75 69 6c  64 72 6f 6f 74 2f 75 73  |cts..uildroot/us|
> 000cacd0  72 2f 6c 69 62 2f 72 61  63 6b 65 74 2f 63 6f 6c  |r/lib/racket/col|
> 000cace0  6c 65 63 74 73 00 00 2a  2a 2a 2a 2a 2a 2a 2a 2a  |lects..*********|
> 000cacf0  2a 2a 2a 2a 2a 2a 2a 2a  2a 2a 2a 2a 2a 2a 2a 2a  |****************|
> *
> ```

These look like the false positives @mflatt mentioned in https://github.com/racket/racket/issues/3726#issuecomment-798734769:

> It's true that the `fix-paths` step can leave a tail of the `DESTDIR` path in the executable, since it only writes the new path and leaves trailing bytes as-is. That could be improved, but it's not new to CS.

(I do think it would be good to improve that, since it would make an otherwise-[reproducible](https://reproducible-builds.org/) build non-reproducible, but it's not the only reproducible builds issues affecting Racket.)

**Comment by mflatt** [comment]
> I was surprised just now to see that building racket-vm-cs with pre-built Chez and Racket BC took 13,670 seconds, whereas the total time up to that point was 8,301, including for bootstrapping Chez via CGC. Is that sort of timing expected, maybe due to WPO or something? 

It might be the time to compile "pbchunk" C code that is generated from bytecode. That's the sequence in the log compiling "petite0.c" through "racket9.c". I've put some effort into making the generated C more compiler-friendly, but the generated files are still large and weird-looking to a C compiler.

**Comment by LiberalArtist** [comment]
> > I was surprised just now to see that building racket-vm-cs with pre-built Chez and Racket BC took 13,670 seconds, whereas the total time up to that point was 8,301, including for bootstrapping Chez via CGC. Is that sort of timing expected, maybe due to WPO or something?
> 
> It might be the time to compile "pbchunk" C code that is generated from bytecode. That's the sequence in the log compiling "petite0.c" through "racket9.c". I've put some effort into making the generated C more compiler-friendly, but the generated files are still large and weird-looking to a C compiler.

That makes sense, and it may explain some of the difference even if https://github.com/racket/racket/issues/4492 is also a factor.

Does building Chez directly not use "pbchunk" mode, then? Should it?

I'll try to look into this a bit more after I've got the 8.7 release into Guix.

**Comment by Ancieg** [comment]
> > > Perhaps try `-j 1` to disable parallel jobs.
> > 
> > 
> > I will try using `-j 1`, but I am not really sure it will decrease CPU-time eating (our builder has limited value of `ulimit -t`).
> 
> That helps building on `i586` only (so, now I get builds on `x86_64`, `aarch64` and `i586`). Builds on `ppc64le` and `armh` still crash due to `CPU time limit exceeded`.

I temporary turned off building main-distribution packages for `i586`, `armh` and `ppc64le` :disappointed: (see [308872](https://git.altlinux.org/tasks/308872/)).
Now I am going to update racket up to 8.7.
I think the issue should be closed, because I have no idea when I be ready to contribute to the project because I am going to learn Racket for the near future.
